src/Microsoft.ML.Data/Data/ITransposeDataView.cs

```diff
@@ -14,46 +14,30 @@ namespace Microsoft.ML.Data
+    /// (naturally, as opposed to row-by-row through an <see cref="RowCursor"/>). This interface
+    /// is intended to be implemented by classes that want to provide an option for an alternate
+    /// way of accessing the data stored in a <see cref="IDataView"/>.
     ///
-    /// The interface only advertises that columns may be accessible in slot-wise fashion. A column
-    /// is accessible in this fashion iff <see cref="TransposeSchema"/>'s
-    /// <see cref="ITransposeSchema.GetSlotType"/> returns a non-null value.
+    /// The interface only advertises that columns may be accessible in slot-wise fashion. The i-th column
+    /// is accessible in this fashion iff <see cref="TransposeSlotTypes"/>[i] is not <see langword="null"/>
+    /// and <see cref="TransposeSlotTypes"/>[i] is not <see langword="null"/>.
     /// </summary>
     [BestFriend]
     internal interface ITransposeDataView : IDataView
     {
-        /// <summary>
-        /// An enhanced schema, containing information on the transposition properties, if any,
-        /// of each column. Note that there is no contract or suggestion that this property
-        /// should be equal to <see cref="IDataView.Schema"/>.
-        /// </summary>
-        ITransposeSchema TransposeSchema { get; }
-
         /// <summary>
         /// Presents a cursor over the slots of a transposable column, or throws if the column
         /// is not transposable.
         /// </summary>
         SlotCursor GetSlotCursor(int col);
-    }
 
-    /// <summary>
-    /// The transpose schema returns the schema information of the view we have transposed.
-    /// </summary>
-    [BestFriend]
-    internal interface ITransposeSchema : ISchema
-    {
         /// <summary>
-        /// Analogous to <see cref="ISchema.GetColumnType"/>, except instead of returning the type of value
-        /// accessible through the <see cref="RowCursor"/>, returns the item type of value accessible
-        /// through the <see cref="SlotCursor"/>. This will return <c>null</c> iff this particular
-        /// column is not transposable, that is, it cannot be viewed in a slotwise fashion. Observe from
-        /// the return type that this will always be a vector type. This vector type should be of fixed
-        /// size and one dimension.
+        /// <see cref="TransposeSlotTypes"/>[i] specifies the type of all values at the i-th column of <see cref="IDataView"/>.
+        /// For example, if <see cref="IDataView.Schema"/>[i] is a scalar float column, then <see cref="TransposeSlotTypes"/>[i]
+        /// may return a <see cref="VectorType"/> whose <see cref="VectorType.ItemType"/> field is <see cref="NumberType.R4"/>.
+        /// If the i-th column can't be iterated column-wisely, <see cref="TransposeSlotTypes"/>[i] may be <see langword="null"/>.
         /// </summary>
-        VectorType GetSlotType(int col);
+        VectorType[] TransposeSlotTypes { get; }
```

Comment:
>VectorType[] [](start = 8, length = 12)

I would prefer the old method, if possible. However, if you insist on replacing it with a collection, I'd prefer that it be something not mutable. Like, say, a read only list, or an `System.Collections.Immutable` array, or something like this. Having what is essentially part of the schema be a mutable structure is not an acceptable state even for something internal.

Reply:
Your wish is granted.

---
In reply to: [248777208](https://github.com/dotnet/machinelearning/pull/2111#discussion_r248777208) [](ancestors = 248777208)

Reply:
Cool, I have two left I think!

---
In reply to: [248815747](https://github.com/dotnet/machinelearning/pull/2111#discussion_r248815747) [](ancestors = 248815747,248777208)